File: javascript/space-age/space-age.js

const SECONDS_IN_EARTH_YEAR = 365.25 * 24 * 60 * 60

// Number of "Earth years" for each planet's full orbit.
const ORBITAL_PERIODS = {
  'mercury': 0.2408467,
  'venus': 0.61519726,
  'earth': 1,
  'mars': 1.8808158,
  'jupiter': 11.862615,
  'saturn': 29.447498,
  'uranus': 84.016846,
  'neptune': 164.79132,
}

export const age = (planet, seconds) => Number((seconds / SECONDS_IN_EARTH_YEAR / ORBITAL_PERIODS[planet]).toFixed(2))

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet (..), ageOn) where

data Planet = Mercury | Venus | Earth | Mars | Jupiter | Saturn| Uranus | Neptune

secondsPerEarthYear :: Int
secondsPerEarthYear = 31557600

earthYearsForPlanet :: Fractional p => Planet -> p
earthYearsForPlanet planet = case planet of
  Earth -> 1
  Mercury -> 0.2408467
  Venus -> 0.61519726
  Mars -> 1.8808158
  Jupiter -> 11.862615
  Saturn -> 29.447498
  Uranus -> 84.016846
  Neptune -> 164.79132

earthYearsFromSeconds :: Float -> Float
earthYearsFromSeconds seconds = seconds / fromIntegral secondsPerEarthYear

ageOn :: Planet -> Float -> Float
ageOn planet seconds = earthYearsFromSeconds seconds / earthYearsForPlanet planet
